sre_maturity.md

```diff
@@ -0,0 +1,115 @@
+
+Following on from this, you have an escalation process. A way to be notified of problems. 
+It should be possible to raise a problem with the SRE team and the people that build the service, like engineers, QE, writers.
+This can be a mailing list, slack channel, issue system, whatever.
+Formalise it and share it.
+
+If things go really bad, you write up an RCA.
+You should be starting to build up a library of RCAs that can feed into doing more proactive things as you mature e.g. Incident rehearsals.
+Each RCA includes a 'blameless' post mortem session with the people involved while the incident is still fresh in their head.
+This post mortem feeds important reliability improvements back to the team.
+
+#### Room to improve
+
+Your team has sufficient time in their day to pursue improvement.
+Your team needs room to practice processes, improve them and automate things.
+Without this room, your team will mature more slowly and get bogged down in just keeping the service running day to day.
+This can lead to scaling problems, both for the service and the team.
+It isn't necessary to formally measure how much time is spent on keeping things running vs. feature development or automation.
+That can come later, however you are familiar with the concept of Toil.
+
+
+## Phase 2: The ‘Walk’ Phase (WIP)
+
+### Overview
+
+Your team is well established, has working processes and is familiar with SRE principles.
+A lot of those principles are baked into their processes & daily work.
+In this phase your team is building on the foundations of observability and starting to reflect on their learnings about the service. 
+Feedback loops, automation and 'shifting left' with different types of testing are the main theme of this phase.
+Your team is capable of scaling the service and number of users in a profitable way.
+
+### Characteristics
+
+* SLOs - SLOs have been iterated on, and you have more than 1 meaningful SLO. You understand and track error budgets
+* Observability
+    * Actionable Alerts
+    * Self healing
+        * Avoid SRE having sub-systems that automagically work around the inefficiencies of upstream - needs a good argument why it can’t be fixed upstream - pushing left
+    * Combination of internal (whitebox) and external (blackbox) monitoring - tied to SLOs
+    * Dashboards, logging (easily accessible)
+* Chaos Testing (possible candidate for Phase 1, but what would that look like? Ad-hoc testing?)
+* Starting to get Exec support (improvment on 'buy-in')
+* Start measuring Toil more formally & management track it
+    * Develop relationship with the upstream (eng and sre, or sre & upstream product) for managing toil
+* Improving the architecture for lower cost to serve (Is there a prereq here - calculating & knowing the cost to serve)
+* DR & Incident Rehearsals (Identifying failure modes could be a prerequisite in phase 1)
+
```

Comment:
Could we add something here about "appropriate error logging". IMO, there is a habit used by devs to "chuck" all sorts of stuff into the logs (the old days of printf your output to debug!). They use it to help their development. I know because I have done this myself as a very foundational method :). This is fine for the first phase, but the remnants of these can hang around, even years later.
We have a mantra of: "Errors should actually be errors". The logs are the "UI" for SREs for the health of a service. If the logs are "polluted" by a whole lot of other non-errors, then that just makes it harder to resolve errors and increases the MTTR. I would also include in this the problem of not having errors where there should be an error.
Ideally, SRE is only notified when the actually is an error - but I'd say that is more for "phase 3".

Reply:
@brizrobbo Yes, definitely.
I'm thinking there's a lot there that could be captured as 'best practices' and linked off from the maturity model under the 'observability' headings in each phase.
As it stands I think https://issues.redhat.com/browse/SIGSRE-54 is where the logging best practices would be captured in such a doc (that issue may split as there's a lot in 'best practices').

I'm wondering how best to keep it sufficiently high level in these maturity phases and link off to additional info, guides, templates & worked examples where it makes sense.
In this case, I think saying something like this for phase 2?

```
You have established patterns and guidelines for observability signals such as metrics, logs, errors etc...
You are using these patterns to speed up debugging of issues and correlating between signals easier, thereby reducing your Mean Time To Recovery (MTTR).
Your alerts have matured and rarely give false positives as a result of having these patterns and refining queries.
```
How does that sound?

Phase 3 would then build on this foundation looking at AI Ops and improving the debug experience (focused on decreasing MTTR further)

Reply:
The topic of logging also makes sense to mention in the operable software piece of the SIG's interests.